### DISCUSSION TITLE
Possible improvements/changes in the CP-SAT code

### DISCUSSION THREAD

**Discussion by anubhav-cs** [google/or-tools#3534] [CP-SAT questions] [answered]
While extending CP-SAT for my research I changed a few lines of code. I just want to make sure that there are no apparent flaws in them, hence posting this.

  1.  In the expansion of Variable Elements constraints, there is a `BoolOr` constraint over the literals in the index variable encoding. I believe this could be replaced by `ExactlyOne`, since an integer variable takes exactly one value, hence exactly one literal can be true. 
  https://github.com/google/or-tools/blob/82750ac12f1ee5354e1c7869894d9af3508778f2/ortools/sat/cp_model_expand.cc#L471
  
  2. The `backtrack` method updates the current decision level and target trail index by iteratively decreasing the current decision level until it equals the target level. I think this could be replaced with these two lines.
  https://github.com/google/or-tools/blob/82750ac12f1ee5354e1c7869894d9af3508778f2/ortools/sat/sat_solver.cc#L1024
         
          current_decision_level_ = target_level;
          int target_trail_index =  decisions_[current_decision_level_].trail_index;

_Summary of changes_  https://github.com/google/or-tools/compare/stable...anubhav-cs:or-tools:stable

Thanks in advance :)

**Comment by lperron** [comment]
We tried that for the element. It does not helpand exactly one is much slower than a boolor.

**Reply by anubhav-cs** [reply]
Thanks for the reply! I have a followup question. Is there a way to keep the variable indexes in CP model proto intact without setting `keep_all_feasible_solutions_in_presolve` `ON`?

I am writing custom propagators where I need the same variable indexes as the client-side model proto. I could achieve this by setting `keep_all_feasible_solutions_in_presolve` `ON`, but it increases the solve time of the vanilla model. Keeping all feasible solutions seems unnecessary. In fact, things like symmetry breaking during presolve, which are disabled by this setting, are nice to keep.

**Reply by anubhav-cs** [reply]
Additionally, going by the code below, `exactly_one` constraint does get applied to the fully encoded variables in certain scenarios, given that `keep_all_feasible_solutions_in_presolve` is `OFF`. Index variables are fully encoded, so this applies to them in my model.
https://github.com/google/or-tools/blob/82750ac12f1ee5354e1c7869894d9af3508778f2/ortools/sat/cp_model_presolve.cc#L8305

After doing some experiments, I realized that the exactly_one constraint applied in the above step was actually helping. Hence, the change mentioned in the initial message.

**Comment by lperron** [comment]
Thanks. I will integrate both changes in our codebase.

**Comment by lperron** [comment, marked as answer]
And done.

Thanks
